src/main/java/com/mariamkatamashvlii/gym/service/implementation/TrainerServiceImpl.java

```diff
@@ -0,0 +1,194 @@
+            Trainer trainer = Trainer.builder()
+                    .specialization(type)
+                    .user(user)
+                    .build();
+            trainerRepo.save(trainer);
+            log.info("[{}] Registered new trainer with username: {}", transactionId, user.getUsername());
+            return new RegistrationResponseDTO(user.getUsername(), user.getPassword());
+        } catch (EntityNotFoundException e) {
+            log.error("[{}] Error during registration - {}", transactionId, e.getMessage());
+            throw new ResponseStatusException(HttpStatus.NOT_FOUND, e.getMessage(), e);
+        } catch (Exception e) {
+            log.error("[{}] Unexpected error during trainer creation: {}", transactionId, e.getMessage());
+            throw new UserNotCreatedException("Could not create trainer due to an unexpected error");
+        }
+    }
+
+    @Override
+    public ProfileResponseDTO getTrainerProfile(String username) {
+        String transactionId = UUID.randomUUID().toString();
+        log.info("[{}] Fetching profile for username: {}", transactionId, username);
+
+        Trainer trainer = trainerRepo.findByUsername(username);
+        if (trainer == null) {
+            log.error("[{}] Trainer not found with username: {}", transactionId, username);
+            throw new EntityNotFoundException("Trainer not found with username - " + username);
+        }
+
+        TrainingTypeDTO specialization = new TrainingTypeDTO(
+                trainer.getSpecialization().getId(),
+                trainer.getSpecialization().getTrainingTypeName()
+        );
+
+        List<TraineeDTO> trainees = trainer.getTrainees().stream().map(trainee -> new TraineeDTO(
+                trainee.getUser().getUsername(),
+                trainee.getUser().getFirstName(),
+                trainee.getUser().getLastName()
+        )).toList();
+
+        log.info("[{}] Successfully fetched profile for username: {}", transactionId, username);
+        return new ProfileResponseDTO(
+                trainer.getUser().getFirstName(),
+                trainer.getUser().getLastName(),
+                specialization,
+                trainer.getUser().getIsActive(),
+                trainees
+        );
+    }
+
+    @Override
+    public UpdateResponseDTO updateProfile(UpdateRequestDTO updateRequestDTO) {
+        String transactionId = UUID.randomUUID().toString();
+        String username = updateRequestDTO.getUsername();
+        log.info("[{}] Attempting to update profile for username: {}", transactionId, username);
+
+        User user = userRepo.findByUsername(username);
+        if (user == null) {
+            log.error("[{}] User not found with username: {}", transactionId, username);
+            throw new EntityNotFoundException("User not found for username: " + username);
+        }
```

Comment:
Here I'd suggest the foolowing: introduce component for validation operations. F.e. TrainerValidator. Its responsibility will inlude all the verifications and throwing exceptions.
It will make service more clear and readable.

In addition, please follow the method name as it says what it does. updateProfile should just update the profile and not verify anything.